File: js/spec/gilded_rose_spec.js
describe('Gilded Rose', function() {
  it('should foo', function() {
    const gildedRose = new Shop([new Item('foo', 0, 0)]);
    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual('foo');
  });

  it('can updateQuality for Backstage Passes', () => {
    const gildedRose = new Shop([
      new Item('Backstage passes to a TAFKAL80ETC concert', 15, 20),
      new Item('Backstage passes to a TAFKAL80ETC concert', 10, 46),
      new Item('Backstage passes to a TAFKAL80ETC concert', 5, 49)
    ]);
    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual('Backstage passes to a TAFKAL80ETC concert');
    expect(items[0].sellIn).toEqual(14);
    expect(items[0].quality).toEqual(21);

    expect(items[1].name).toEqual('Backstage passes to a TAFKAL80ETC concert');
    expect(items[1].sellIn).toEqual(9);
    expect(items[1].quality).toEqual(48);

    expect(items[2].name).toEqual('Backstage passes to a TAFKAL80ETC concert');
    expect(items[2].sellIn).toEqual(4);
    expect(items[2].quality).toEqual(50);
  });

  it('can updateQuality for Aged Brie', () => {
    const gildedRose = new Shop([new Item('Aged Brie', 2, 0)]);
    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual('Aged Brie');
    expect(items[0].sellIn).toEqual(1);
    expect(items[0].quality).toEqual(1);
  });

  it('can updateQuality for Sulfuras', () => {
    const gildedRose = new Shop([
      new Item('Sulfuras, Hand of Ragnaros', 0, 80),
      new Item('Sulfuras, Hand of Ragnaros', -1, 80)
    ]);
    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual('Sulfuras, Hand of Ragnaros');
    expect(items[0].sellIn).toEqual(-1);
    expect(items[0].quality).toEqual(80);

    expect(items[1].name).toEqual('Sulfuras, Hand of Ragnaros');
    expect(items[1].sellIn).toEqual(-2);
    expect(items[1].quality).toEqual(80);
  });

  it('can updateQuality for basic items', () => {
    const gildedRose = new Shop([
      new Item('+5 Dexterity Vest', 10, 20),
      new Item('Elixir of the Mongoose', 5, 7)
    ]);

    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual('+5 Dexterity Vest');
    expect(items[0].sellIn).toEqual(9);
    expect(items[0].quality).toEqual(19);

    expect(items[1].name).toEqual('Elixir of the Mongoose');
    expect(items[1].sellIn).toEqual(4);
    expect(items[1].quality).toEqual(6);
  });

  it('can updateQuality for conjured items', () => {
    const gildedRose = new Shop([new Item('Conjured Mana Cake', 3, 6)]);
    const items = gildedRose.updateQuality();
    expect(items[0].sellIn).toEqual(2);
    expect(items[0].quality).toEqual(4);
  });

  it('Once the sell by date has passed, Quality degrades twice as fast', () => {
    const gildedRose = new Shop([new Item('+5 Dexterity Vest', -2, 20)]);

    const items = gildedRose.updateQuality();
    expect(items[0].sellIn).toEqual(-3);
    expect(items[0].quality).toEqual(18);
  });

  it('The Quality of an item is never negative', () => {
    const gildedRose = new Shop([new Item('+5 Dexterity Vest', -2, 1)]);

    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual('+5 Dexterity Vest');
    expect(items[0].sellIn).toEqual(-3);
    expect(items[0].quality).toEqual(0);
  });

  it('will not increase quality over 50', () => {
    const gildedRose = new Shop([new Item('Aged Brie', -1, 50)]);
    const items = gildedRose.updateQuality();
    expect(items[0].name).toEqual('Aged Brie');
    expect(items[0].sellIn).toEqual(-2);
    expect(items[0].quality).toEqual(50);
  });
});

File: ruby/spec/gilded_rose_spec.rb
require 'gilded_rose'

describe GildedRose do
let(:potato) { Item.new('potato', 15, 2)}
let(:sulfarus) { Item.new('Sulfuras, Hand of Ragnaros', 50, 80) }

  describe "feature tests" do
    it "does not change the name" do
      items = [Item.new("foo", 0, 0)]
      GildedRose.update_quality(items)
      expect(items[0].name).to eq "foo"
    end
    describe 'non-special item input' do
      it 'should decrese the quality of normal item' do
        items = [potato]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq (1)
      end

      it 'should decrese the sellIn of normal item' do
        items = [potato]
        GildedRose.update_quality(items)
        expect(items.first.sell_in).to eq (14)
      end

      it 'should decrease quality of normal items by 2 when sell_in date passes' do
        items = [Item.new("old potato", -1, 20)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq (18)
      end

      it 'should not decrease quality below 0' do
        items = [Item.new("old potato", 0, 0)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq (0)
      end
    end

    

    describe 'Aged Brie input' do
      it 'increases in quality as it ages' do
        items = [Item.new("Aged Brie", 25, 45)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 46
      end

      it 'cannot increase in quality beyond 50' do
        items = [Item.new("Aged Brie", 25, 50)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 50
      end
      it 'doubles the increase when sell_in below 1' do
        items = [Item.new("Aged Brie", 0, 20)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 22
      end
    end

    describe 'Sulfuras input' do
      it 'does not change' do
        items = [sulfarus]
        GildedRose.update_quality(items)
        expect(items[0].quality).to eq 80
        expect(items[0].sell_in).to eq 50
      end
    end

    describe 'backstage pass input' do
      it 'increases in quality by one when sellIn is > 10 days' do
        items = [Item.new("Backstage passes to a TAFKAL80ETC concert", 15, 20)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 21
      end

      it 'increases in quality by 2 when sellin < 10 days' do
        items = [Item.new("Backstage passes to a TAFKAL80ETC concert", 9, 20)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 22
      end

      it 'increases by 3 when sellin < 5' do
        items = [Item.new("Backstage passes to a TAFKAL80ETC concert", 4, 20)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 23
      end

      it 'decreases to 0 when sellin == 0' do
        items = [Item.new("Backstage passes to a TAFKAL80ETC concert", 0, 20)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 0
      end
    end

    describe 'conjured input' do
      it "should degrade twice as fast as normal items" do
        items = [Item.new("Conjured Armour", 10, 15)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 13
      end
      
      it "should not degrade past 0" do
        items = [Item.new("Conjured Armour", 10, 1)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 0
      end

      it "should degrade by 4 when sell in has passed" do
          items = [Item.new("Conjured Armour", -1, 17)]
          GildedRose.update_quality(items)
          expect(items.first.quality).to eq 13
      end

      it "still should not degrade past 0 when sell in has passed" do
        items = [Item.new("Conjured Armour", -1, 3)]
        GildedRose.update_quality(items)
        expect(items.first.quality).to eq 0
      end 
    end


    
 
end
  describe '#update_normal_quality' do
    it 'updates the quality of a normal item' do
      item_double = double :item, name: "potato", sell_in: 1, quality: 3

      expect(item_double).to receive(:quality=).with(2)
      GildedRose.update_normal_quality(item_double)
    end

    it 'does not update quality if quality == 0' do
      item_double = double :item, name: "potato", sell_in: 1, quality: 0
      expect(item_double).not_to receive(:quality=).with(-1)
      GildedRose.update_normal_quality(item_double)
    end

  end

  describe '#selfarus?' do
    it 'returns true on a selfarus item' do
      sulfuras_double = double :selfarus, name: 'Sulfuras', sell_in: 50, quality: 80
      expect(GildedRose.sulfuras?(sulfuras_double)).to eq true
    end

    it 'returns false on a non sulfuras item' do
      item_double = double :item, name: "potato", sell_in: 1, quality: 0
      expect(GildedRose.sulfuras?(item_double)).to eq false
    end
  end

  describe '#special_item?' do
    it 'returns true for aged brie' do
      item_double = double :item, name: "Aged Brie"
      expect(GildedRose.special_item?(item_double)).to eq true
    end

    it 'returns true for backstage passes' do
      item_double = double :item, name: "Backstage passes to a TAFKAL80ETC concert"
      expect(GildedRose.special_item?(item_double)).to eq true
    end

    it 'returns true for sulfuras' do
      sulfuras_double = double :selfarus, name: 'Sulfuras'
      expect(GildedRose.special_item?(sulfuras_double)).to eq true
    end 

    it 'returns false on potato' do
      item_double = double :item, name: "potato"
      expect(GildedRose.special_item?(item_double)).to eq false
    end
  end

  describe '#update_backstage_quality' do
     it 'increases in quality by one when sellIn is > 10 days' do
      items_double = double :item, name: "Backstage passes to a TAFKAL80ETC concert", sell_in: 15, quality: 20
      expect(items_double).to receive(:quality=).with(21)
      GildedRose.update_backstage_quality(items_double)
      end

      it 'increases in quality by 2 when sellin < 10 days' do
        items_double = double :item, name: "Backstage passes to a TAFKAL80ETC concert", sell_in: 9, quality: 20
        expect(items_double).to receive(:quality=).with(21).twice
        GildedRose.update_backstage_quality(items_double)
        
      end

      it 'increases by 3 when sellin < 5' do
        items_double = double :item, name: "Backstage passes to a TAFKAL80ETC concert", sell_in: 4, quality: 20
        expect(items_double).to receive(:quality=).with(21).exactly(3).times
        GildedRose.update_backstage_quality(items_double)
      end

      it 'decreases to 0 when sellin == 0' do
        items_double = double :item, name: "Backstage passes to a TAFKAL80ETC concert", sell_in: 0, quality: 20
        expect(items_double).to receive(:quality=).with(0)
        GildedRose.update_backstage_quality(items_double)
      end
  end

  describe '#update_brie_quality' do
    it 'should add one to the quality each day ' do
      brie_double = double :item, name: "Aged Brie", sell_in: 30, quality:15
      expect(brie_double).to receive(:quality=).with(16)
      GildedRose.update_brie_quality(brie_double)
    end

    it 'should not let quality go past 50' do
      brie_double = double :item, name: "Aged Brie", sell_in: 30, quality:50
      expect(brie_double).not_to receive(:quality=).with(51)
      GildedRose.update_brie_quality(brie_double)
    end
    it 'should increase by two when sellin <= 0' do
      brie_double = double :item, name: "Aged Brie", sell_in: 0, quality:2
      expect(brie_double).to receive(:quality=).with(4)
      GildedRose.update_brie_quality(brie_double)
    end


  end

  describe '#brie?' do
    it 'returns true on a brie item' do
      brie_double = double :brie, name: 'Aged Brie', sell_in: 50, quality: 80
      expect(GildedRose.brie?(brie_double)).to eq true
    end

    it 'returns false on a non brie item' do
      item_double = double :item, name: "potato", sell_in: 1, quality: 0
      expect(GildedRose.sulfuras?(item_double)).to eq false
    end
  end

  describe '#backstage?' do
    it 'returns true on a backstage item' do
      backstage_double = double :item, name: "Backstage passes to a TAFKAL80ETC concert"
      expect(GildedRose.backstage?(backstage_double)).to eq true
    end

    it 'returns false on a non brie item' do
      item_double = double :item, name: "potato", sell_in: 1, quality: 0
      expect(GildedRose.backstage?(item_double)).to eq false
    end
  end

 

  

end
